collectors/apicleafcapacity.py

```diff
@@ -0,0 +1,97 @@
+
+        for leaf in data['imdata']:
+            if leaf['eqptcapacityEntity']['children']:
+                """Ignore spine switches without children data"""
+                leaf_data = leaf['eqptcapacityEntity']['children']
+                leaf_dn = leaf['eqptcapacityEntity']['attributes'].split('/')
+                leaf_id = leaf_dn[2]
+            for data_object in leaf_data:
+                if data_object['eqptcapacityL3TotalUsageCap5min']:
+                    l3_tcam_cap_max = data_object['eqptcapacityL3TotalUsageCap5min']['attributes']['v4TotalEpCapMax']
+                elif data_object['eqptcapacityL3TotalUsage5min']:
+                    l3_tcam_cap_used_total = data_object['eqptcapacityL3TotalUsage5min']['attributes']['v4TotalEpAvg']
+                elif data_object['eqptcapacityL3Usage5min']:
+                    l3_tcam_cap_used_local = data_object['eqptcapacityL3Usage5min']['attributes']['v4LocalEpAvg']
+                elif data_object['eqptcapacityL3RemoteUsageCap5min']:
+                    l3_tcam_cap_used_remote = data_object['eqptcapacityL3RemoteUsageCap5min']['attributes']['v4RemoteEpCapAvg']
+                elif data_object['eqptcapacityL2TotalUsage5min']:
+                    l2_tcam_cap_max = data_object['eqptcapacityL2TotalUsage5min']['attributes']['totalEpCapMax']
+                    l2_tcam_cap_used_total = data_object['eqptcapacityL2TotalUsage5min']['attributes']['totalEpAvg']
+                elif data_object['eqptcapacityL2Usage5min']:
+                    l2_tcam_cap_used_local = data_object['eqptcapacityL2Usage5min']['attributes']['localEpAvg']
+                elif data_object['eqptcapacityL2RemoteUsage5min']:
+                    l2_tcam_cap_used_remote = data_object['eqptcapacityL2RemoteUsage5min']['attributes']['remoteEpAvg']
+            g_leaf_cap_tcam_l3_max.add_metric(labels=[leaf_id, 'max', 'l3'],
+                                value=l3_tcam_cap_max)
+            g_leaf_cap_tcam_l3_used_total.add_metric(labels=[leaf_id, 'total', 'l3'],
+                                value=l3_tcam_cap_used_total)
+            g_leaf_cap_tcam_l3_used_local.add_metric(labels=[leaf_id, 'local', 'l3'],
+                                value=l3_tcam_cap_used_local)
+            g_leaf_cap_tcam_l3_used_remote.add_metric(labels=[leaf_id, 'remote', 'l3'],
+                                value=l3_tcam_cap_used_remote)
+            g_leaf_cap_tcam_l2_max.add_metric(labels=[leaf_id, 'max', 'l2'],
+                                value=l2_tcam_cap_max)
+            g_leaf_cap_tcam_l2_used_total.add_metric(labels=[leaf_id, 'total', 'l2'],
+                                value=l2_tcam_cap_used_total)
+            g_leaf_cap_tcam_l2_used_local.add_metric(labels=[leaf_id, 'local', 'l2'],
+                                value=l2_tcam_cap_used_local)
+            g_leaf_cap_tcam_l2_used_remote.add_metric(labels=[leaf_id, 'remote', 'l2'],
+                                value=l2_tcam_cap_used_remote)
```

Comment:
I propose merging the the retrieval of the metric value and the creation of the metric. As we already know the label values for `layer` and `usage` and can add the metric if it exists.
From letting this run locally I have seen that a leaf with a usage of 0 will also return 0's, so we should not have any issue with missing values.
```suggestion
                if 'eqptcapacityL3TotalUsageCap5min' in data_object:
                    l3_max = data_object['eqptcapacityL3TotalUsageCap5min']['attributes']['v4TotalEpCapMax']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'max', 'l3'],
                                               value=l3_max)
                elif 'eqptcapacityL3TotalUsage5min' in data_object:
                    l3_total = data_object['eqptcapacityL3TotalUsage5min']['attributes']['v4TotalEpAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'total', 'l3'],
                                               value=l3_total)
                elif 'eqptcapacityL3Usage5min' in data_object:
                    l3_local = data_object['eqptcapacityL3Usage5min']['attributes']['v4LocalEpAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'local', 'l3'],
                                               value=l3_local)
                elif 'eqptcapacityL3RemoteUsageCap5min' in data_object:
                    l3_remote = data_object['eqptcapacityL3RemoteUsageCap5min']['attributes']['v4RemoteEpCapAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'remote', 'l3'],
                                               value=l3_remote)
                elif 'eqptcapacityL2TotalUsage5min' in data_object:
                    l2_max = data_object['eqptcapacityL2TotalUsage5min']['attributes']['totalEpCapMax']
                    l2_total = data_object['eqptcapacityL2TotalUsage5min']['attributes']['totalEpAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'max', 'l2'],
                                               value=l2_max)
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'total', 'l2'],
                                               value=l2_total)
                elif 'eqptcapacityL2Usage5min' in data_object:
                    l2_local = data_object['eqptcapacityL2Usage5min']['attributes']['localEpAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'local', 'l2'],
                                               value=l2_local)
                elif 'eqptcapacityL2RemoteUsage5min' in data_object:
                    l2_remote = data_object['eqptcapacityL2RemoteUsage5min']['attributes']['remoteEpAvg']
                    g_leaf_cap_tcam.add_metric(labels=[leaf_id, 'remote', 'l2'],
                                               value=l2_remote)
```